### DISCUSSION TITLE
Exchange communicaiton model

### DISCUSSION THREAD

**Discussion by cad-x2** [trinodb/trino#13116] [General]
I'm trying to understand the communication model between the exchange client and the output buffer in workers.
As far as I understand the exchange client on downstream workers pull pages directly from the output buffers of upstream workers. But how is the signalling done? Is the exchange client periodically checking for pages or is a signal sent from the upstream node indicating that there is data available
![v2-14c069c25672b445236a93c02d7a4125_720w](https://user-images.githubusercontent.com/23506961/177808883-040feb77-33d4-43c8-9a8f-085aab91c7b9.png)
?

**Comment by electrum** [comment]
The exchange, like several other things in Trino, uses HTTP long polling. The client makes a request, and the server waits for some period of time before returning the response if there is no data available.

This is implemented using JAX-RS `@Suspended AsyncResponse`. The resource registers a future that will complete when data is available. The request is then resumed and the data is returned to the client. Using async means that these long polling responses don't block server threads.
